### ISSUE TITLE
Question. Problems with the launch of Podman in ARM64 emulated. 

### ISSUE THREAD

**Issue by AleksMosk** [mgoltzsche/podman-static#13]
I built podman image for ARM64 using your project and qemu(I added ARM64V8 / prefix for basic images to DockerFile) using these commands:

`$docker run --rm --privileged multiarch/qemu-user-static --reset -p yes`
`$docker build --force-rm --platform=linux/arm64/v8 -t arm64-podman .`

After successful building, I launched a container:
`#docker run --privileged -it  mgoltzsche/podman`

Inside the container, I tried to test several commands:
    `#podman --version` returned:
```
ERRO[0000] failure getting variant error="getCPUInfo for pattern: Cpu architecture: not found"  
podman version 3.2.1
```
``
    `#podman -info`, `#podman run` and others returned the same error as in https://github.com/containers/podman/issues/9164:
```
ERRO[0000] failure getting variant error="getCPUInfo for pattern: Cpu architecture: not found"  
Error: failed to get new shm lock manager: failed to create 2048 locks in /libpod_lock: function not implemented
```

Do I need any changes in the configuration files? Maybe I build an image with incorrect flags, or incorrectly start it?

**Comment by mgoltzsche** [comment]
Interesting!
To really make sure that the image with the correct architecture is used I think one should use the `--pull` option with the `docker build` command to make sure you're not using previously loaded amd64 images.
However even with that option enabled I can reproduce your problem on my machine.

At least the "getCPUInfo" error seems to be coming from [here](https://github.com/containers/image/blob/v5.12.0/internal/pkg/platform/platform_matcher.go#L62) where an architecture detection function searches for a `Cpu architecture` attribute within `/proc/cpuinfo` which is present on arm systems apparently but not in the emulated docker container. Also see https://github.com/docker/for-linux/issues/1146.

**Comment by AleksMosk** [comment]
Thanks for your answer.
Architecture is correct.
I just checked my image using `docker image inspect arm64/podman:latest`.
It returned ` "Architecture": "arm64",`.
The full result is below:
```
{
    "Id": "sha256:3239562b64e9bf946f8b2c6b315c6c46b3015b4add51ef91a279a995a903d3b3",
    "RepoTags": [
        "arm64/podman:latest"
    ],
    "RepoDigests": [],
    "Parent": "sha256:41214e94f635b4529ae346f033670fee64e5b3a314935bd488459c4098c8a66f",
    "Comment": "",
    "Created": "2021-07-08T09:06:27.124418441Z",
    "Container": "",
    "ContainerConfig": {
        "Hostname": "",
        "Domainname": "",
        "User": "",
        "AttachStdin": false,
        "AttachStdout": false,
        "AttachStderr": false,
        "Tty": false,
        "OpenStdin": false,
        "StdinOnce": false,
        "Env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "_CONTAINERS_USERNS_CONFIGURED=",
            "BUILDAH_ISOLATION=chroot",
            "container=oci"
        ],
        "Cmd": [
            "/bin/sh",
            "-c",
            "#(nop) COPY dir:ef61cbaea018c4a2c7b767a96af8ec6d293c3e2f07a7b16de09148451573047f in /etc/cni "
        ],
        "Image": "sha256:41214e94f635b4529ae346f033670fee64e5b3a314935bd488459c4098c8a66f",
        "Volumes": null,
        "WorkingDir": "",
        "Entrypoint": null,
        "OnBuild": null,
        "Labels": null
    },
    "DockerVersion": "20.10.2",
    "Author": "",
    "Config": {
        "Hostname": "",
        "Domainname": "",
        "User": "",
        "AttachStdin": false,
        "AttachStdout": false,
        "AttachStderr": false,
        "Tty": false,
        "OpenStdin": false,
        "StdinOnce": false,
        "Env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "_CONTAINERS_USERNS_CONFIGURED=",
            "BUILDAH_ISOLATION=chroot",
            "container=oci"
        ],
        "Cmd": [
            "/bin/sh"
        ],
        "Image": "sha256:41214e94f635b4529ae346f033670fee64e5b3a314935bd488459c4098c8a66f",
        "Volumes": null,
        "WorkingDir": "",
        "Entrypoint": null,
        "OnBuild": null,
        "Labels": null
    },
    "Architecture": "arm64",
    "Os": "linux",
    "Size": 108797583,
    "VirtualSize": 108797583,
    "GraphDriver": {
        "Data": {
            "LowerDir": "/var/lib/docker/overlay2/dc3c74e8248922698823563db93c72e13558254cad6bf697ab2e844ca940b4dd/diff:/var/lib/docker/overlay2/7b83e1f0dfe61a88ba6085dcbce138aecf336e3533c453dc9e1772545eaeecab/diff:/var/lib/docker/overlay2/6cfa52cdc7e0eaccd64e4ebbd7814b1cb63c465138968be5f9edf4807f505555/diff:/var/lib/docker/overlay2/b46a1c28a2ea8bf73e4431551b742eaf1d7a7ec56fe21ca581030fba8d6f64b6/diff:/var/lib/docker/overlay2/42d92623ded3aa0d2367fb02fdd7e857a59ef6a811ae2a57c74fef782fb256d4/diff:/var/lib/docker/overlay2/6b9af5018aacfe4e75731e78ba8781fbaa8c099d559518e0d6efac2b821c64c4/diff:/var/lib/docker/overlay2/721e2f042536a7799907e138bed9462e7fb1815b0be3c4cd0abd63e2c62ac355/diff:/var/lib/docker/overlay2/72c2e05a81ae5d90dd8f1ccd4b69e1a8e6dfe2bb6bd5c7f448da50618c9507eb/diff:/var/lib/docker/overlay2/f1760f69894faf2fb9dc64a7e0eeef6097451959d2ea60dd760c3a7151e3aa20/diff:/var/lib/docker/overlay2/cb9fa50fe6ca04c495c5d08abdbd25d4db1d10c07be9c3a4c11b39b6f2b747b4/diff:/var/lib/docker/overlay2/49cb832b314eaca1ec00fb3fa9cd478d58b210122e830191076cfb4c9f5e4e6b/diff:/var/lib/docker/overlay2/45eb47bfb83e4c33ab2133cfc8bde15122de85e20616b60c7bcd49dbf6e6ce9e/diff:/var/lib/docker/overlay2/45085364d93499139e10402a31333c88a53ea6c609cae27ddea1d7e2763da337/diff",
            "MergedDir": "/var/lib/docker/overlay2/f61a154f8851f6e5df15fdf677cb21811d5d1797c115bde93780f2a12d14f244/merged",
            "UpperDir": "/var/lib/docker/overlay2/f61a154f8851f6e5df15fdf677cb21811d5d1797c115bde93780f2a12d14f244/diff",
            "WorkDir": "/var/lib/docker/overlay2/f61a154f8851f6e5df15fdf677cb21811d5d1797c115bde93780f2a12d14f244/work"
        },
        "Name": "overlay2"
    },
    "RootFS": {
        "Type": "layers",
        "Layers": [
            "sha256:c55d5dbdab4094da9ba390de49be10dd3b42e990670236a81a792fd2c933fceb",
            "sha256:420b2d921a2845e3866736eb7cb094ae3a5c9ca2f3b0c664cbc0bff145e77699",
            "sha256:659a7d477d6533bf122c08f0abc10d7c11c9931030849afe58b0f1e6301f9efc",
            "sha256:5a23c5d3d63e2f985ede9c3982059ad5a3fc798dc05b9ebc4e787cfd42e13909",
            "sha256:755622a0c7e5f2f3895fa8e1832e1db88f3d5aae2df89bef4e0cf0f190351bfc",
            "sha256:8434fe8f6478c0547834ba5c9322d6cd3910b94b807d73ab49f67efbc44b32ae",
            "sha256:b590d563986e4e06c3809dc829c307255e8fa9fc5da8a778c494feb87bca11ef",
            "sha256:760bdc0d3893d5e9d1ccbf8baaba2e3ae0748f994dcbfa42e63299ab76fccb7f",
            "sha256:7b8718e0ea3370b6a9539a4d1bd352b7ad874a06302f479b18697db66c1b430e",
            "sha256:306f904f8f6fe505437cc407bdea15ba313a62b0cbedca1a11a14c576ffaea75",
            "sha256:28c94568ce32e2a0e682e5bfa1105e092260053a383a2f8bc61ca9318104e2c2",
            "sha256:875e9ce962fa9ab3da95f61811f90693910c03ccf9d3816dece36f142cb64874",
            "sha256:5cfe987923c17495672d7f22f35a467e5b5befd621cda7d6d70425e9d84e50b0",
            "sha256:4f1343130a70d1b7e5cbb66c51e556695393517ac8a742f4200bb9abf7143f73"
        ]
    },
    "Metadata": {
        "LastTagTime": "2021-07-11T21:38:39.561564371+03:00"
    },
    "CreatedTime": 1625735187124,
    "Containers": {}
}
```

I thought the problem is with podman. It returned its version. But when I try to run `docker info` it return `Error: failed to get new shm lock manager: failed to create 2048 locks in /libpod_lock: function not implemented`.

**Comment by AleksMosk** [comment]
Update. 
I changed "lock_type" from "shm" to "file" in the containers.conf then `podman info` returned:
```
host:
  arch: arm64
  buildahVersion: 1.21.0
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - blkio
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - hugetlb
  - pids
  - rdma
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: Unknown
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.0.29, commit: 7e6de6678f6ed8a18661e1d5721b81ccee293b9b'
  cpus: 2
  distribution:
    distribution: alpine
    version: 3.13.5
  eventLogger: file
  hostname: 242e02db6a20
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.4.0-77-generic
  linkmode: dynamic
  memFree: 156413952
  memTotal: 8348930048
  ociRuntime:
    name: runc
    package: Unknown
    path: /usr/local/bin/runc
    version: |-
      runc version 1.0.0-rc95
      commit: b9ee9c6314599f1b4a7f497e1f1f856fe433d3b7
      spec: 1.0.2-dev
      go: go1.16.5
      libseccomp: 2.5.1
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_AUDIT_WRITE,CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_MKNOD,CAP_NET_BIND_SERVICE,CAP_NET_RAW,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: false
    seccompProfilePath: ""
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 2078273536
  swapTotal: 2147479552
  uptime: 46h 50m 53s (Approximately 1.92 days)
registries:
  search:
  - docker.io
  - registry.fedoraproject.org
  - registry.access.redhat.com
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.ignore_chown_errors: "true"
    overlay.mount_program:
      Executable: /usr/local/bin/fuse-overlayfs
      Package: Unknown
      Version: |-
        fuse-overlayfs: version 1.5
        fusermount3 version: 3.10.4
        FUSE library version 3.10.4
        using FUSE kernel interface version 7.31
    overlay.mountopt: nodev,fsync=0
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: overlayfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 0
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.2.1
  Built: 1625733592
  BuiltTime: Thu Jul  8 08:39:52 2021
  GitCommit: 152952fe6b18581615c3efd1fafef2d8142738e8
  GoVersion: go1.16.5
  OsArch: linux/arm64
  Version: 3.2.1

DEBU[0000] Called info.PersistentPostRunE(/usr/local/bin/podman info --log-level=debug)
```

**But** `podman run alpine:latest echo hello` returned: 
```
Error: Error initializing destination containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.ignore_chown_errors=true,overlay.mount_program=/usr/local/bin/fuse-overlayfs,overlay.mountopt=nodev,fsync=0]docker.io/library/alpine:latest: error creating a temporary directory: stat /data/tmp: no such file or directory
```

**Comment by mgoltzsche** [comment]
> I changed "lock_type" from "shm" to "file" in the containers.conf

That makes sense since the upstream arm64 build [disables cgo](https://github.com/containers/podman/blob/v3.2.2/Makefile#L376) although the lock implementation causing the previous error requires cgo.

Regarding your last error: it looks like the `/data` directory is missing and must simply be created. I wonder where that path is configured - seems to be new?!
I noticed that the upstream arm64 podman build also [disables overlay support among other features](https://github.com/containers/podman/blob/v3.2.2/Makefile#L68). Therefore, if other problems appear, some of these build tags might fix them for you.

_(Regarding the `failure getting variant` error I was wrong previously: It seems to be coming from [here](https://github.com/containerd/containerd/blob/v1.5.3/platforms/cpuinfo.go#L100))._

**Comment by mgoltzsche** [comment]
@AleksMosk fyi this repository also ships podman arm64 binaries since version 3.4.0.

Meanwhile I've played around a bit more with arm64 builds now that I am using podman on a Raspberry Pi: Turns out the problems you ran into when running podman within a qemu container as you did were caused by the fact that qemu uses a light-weight kernel that does not support all the features podman needs. However when I run those binaries/images on a Raspberry Pi on Raspbian arm64 everything works fine.
(The last problem you had with the data directory was not related to arm64 but must have been due to some customization you did, I guess.)
Also thanks again for creating this issue - I learned a lot from it.
